clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java

```diff
@@ -0,0 +1,291 @@
+     * @return {@link ReconciliationResult}
+     */
+    ReconciliationResult revoke(Optional<Assignment> assignment, Timer timer) {
+        // For revoke, filter out any partitions in our "current" assignment that aren't in the "target" assignment.
+        PartitionDiffer partitionDiffer = (current, target) -> current
+                .stream()
+                .filter(tp -> !target.contains(tp))
+                .collect(Collectors.toSet());
+
+        // For revocation, we modify the "current" assignment by removing all entries that were not present in the
+        // "target" assignment.
+        AssignmentModifier modifier = Set::removeAll;
+
+        return reconcile(assignment,
+                timer,
+                Operation.REVOKE,
+                partitionDiffer,
+                RevokePartitionsEvent::new,
+                modifier);
+    }
+
+    ReconciliationResult assign(Optional<Assignment> assignment, Timer timer) {
+        PartitionDiffer partitionDiffer = (current, target) -> target
+                .stream()
+                .filter(tp -> !current.contains(tp))
+                .collect(Collectors.toSet());
+
+        // For assignment, we modify the "current" assignment by adding all entries that were present in the
+        // "target" assignment but not in the "current" assignment.
+        AssignmentModifier modifier = Set::addAll;
+
+        return reconcile(assignment,
+                timer,
+                Operation.ASSIGN,
+                partitionDiffer,
+                AssignPartitionsEvent::new,
+                modifier);
+    }
+
+    private ReconciliationResult reconcile(Optional<Assignment> assignment,
+                                           Timer timer,
+                                           Operation operation,
+                                           PartitionDiffer partitionDiffer,
+                                           EventGenerator eventGenerator,
+                                           AssignmentModifier assignmentModifier) {
+        // Check for any outstanding operations first. If a conclusive result has already been reached, return that
+        // before processing any further.
+        Optional<ReconciliationResult> inflightStatus = checkInflightStatus(timer);
+
+        if (inflightStatus.isPresent())
+            return inflightStatus.get();
+
+        // "diff" the two sets of partitions: our "current" assignment and the "target" assignment. The result is
+        // sorted primarily so when the partitions show up in the logs, it's easier for us humans to understand.
+        Set<TopicPartition> diffPartitions = sortPartitions(partitionDiffer.diff(subscriptions.assignedPartitions(),
+                targetPartitions(metadata, assignment)));
+
+        // TODO: should this be here? Do we expect the above to do anything besides move things around in memory?
+        timer.update();
```

Comment:
actually - you already are updating the timer no? It doesn't seem necessary here.